### PR TITLE
Support for GRPC based guestagent communication

### PR DESCRIPTION
**Pull request by balajiv113**
Follow up of #1913 

This is more of a prerequisite to address the performance aspects on enabling inotify

### CODE REVIEW

**Comment by balajiv113** [comment]
@lima-vm/maintainers 
Need some suggestions for maintaining proto files. Currently facing issues that we need to generate before running test & build. 

Some ways i could think of is,
- Maintain a separate repo and release tag generate proto files (We can include as go module from there)
- Expose `make test` and rely on proto-gen as dependency like in build 

First option looks clean but more of a overkill for us as our usecase is very limited. 
With second option we will start to use `make test` instead of current `go test ./...`

**Comment by AkihiroSuda** [comment]
> Need some suggestions for maintaining proto files. Currently facing issues that we need to generate before running test & build.

`*.pb.go` can be just committed to the repo
https://github.com/containerd/containerd/tree/main/api/services/containers/v1

CI has to verify that the committed `*.pb.go` can be reproduced from the source.

File: Makefile [hunk 1]
```diff
@@ -242,6 +244,18 @@ lint:
...
+gen-proto:
+	go install google.golang.org/protobuf/cmd/protoc-gen-go@v1.28
+	go install google.golang.org/grpc/cmd/protoc-gen-go-grpc@v1.2
+	go generate ./...
+
+.PHONY: check-proto
+check-proto:
+	@test -z "$$(git status --short | grep ".pb.go" | tee /dev/stderr)" || \
+    		((git diff $$(find . -name '*.pb.go') | cat) && \
+    		(echo "Please run 'make' when making changes to proto files and check-in the generated file changes" && false))
```
Comment (AkihiroSuda) [inline Makefile:257]:
```suggestion
    		(echo "Please run 'make generate' when making changes to proto files and check-in the generated file changes" && false))
```

File: Makefile [hunk 2]
```diff
@@ -242,6 +244,18 @@ lint:
 clean:
 	rm -rf _output vendor
 
+.PHONY: gen-proto
+gen-proto:
```
Comment (AkihiroSuda) [inline Makefile:248]:
```suggestion
generate:
```

File: Makefile [hunk 3]
```diff
@@ -242,6 +244,18 @@ lint:
...
+	go install google.golang.org/protobuf/cmd/protoc-gen-go@v1.28
+	go install google.golang.org/grpc/cmd/protoc-gen-go-grpc@v1.2
+	go generate ./...
+
+.PHONY: check-proto
+check-proto:
```
Comment (AkihiroSuda) [inline Makefile:254]: Maybe this can be meld into `lint:`

File: Makefile [hunk 4]
```diff
@@ -107,6 +107,8 @@ endif
 
 .PHONY: binaries
 binaries: clean \
+    gen-proto \
+    check-proto \
```
Comment (AkihiroSuda) [inline Makefile:111]: No need to be depended by `binaries`

File: pkg/guestagent/api/gen.go [hunk 5]
```diff
@@ -0,0 +1,2 @@
+//go:generate protoc --go_out=. --go_opt=paths=source_relative --go-grpc_out=. --go-grpc_opt=paths=source_relative lima.proto
```
Comment (AkihiroSuda) [inline pkg/guestagent/api/gen.go:1]:
```suggestion
//go:generate protoc --go_out=. --go_opt=paths=source_relative --go-grpc_out=. --go-grpc_opt=paths=source_relative guestservice.proto
```

File: pkg/hostagent/hostagent.go [hunk 6]
```diff
@@ -673,6 +667,7 @@ func (a *HostAgent) processGuestAgentEvents(ctx context.Context, client guestage
 	if err := client.Events(ctx, onEvent); err != nil {
 		return err
 	}
+	logrus.Info("closed")
```
Comment (AkihiroSuda) [inline pkg/hostagent/hostagent.go:670]: Should be `logrus.Debug`?

File: pkg/guestagent/api/const.go [hunk 7]
```diff
@@ -0,0 +1,12 @@
+package api
+
+import (
+	"net"
+	"strconv"
+)
+
+var IPv4loopback1 = net.IPv4(127, 0, 0, 1)
+
+func (x *IPPort) HostString() string {
```
Comment (AkihiroSuda) [inline pkg/guestagent/api/const.go:10]: Doesn't seem to be "const"

File: pkg/guestagent/api/const.go [hunk 8]
```diff
@@ -0,0 +1,12 @@
+package api
+
+import (
+	"net"
+	"strconv"
+)
+
+var IPv4loopback1 = net.IPv4(127, 0, 0, 1)
```
Comment (AkihiroSuda) [inline pkg/guestagent/api/const.go:8]: Technically this is not a const, although I get the point.

File: .github/workflows/release.yml [hunk 9]
```diff
@@ -31,6 +31,7 @@ jobs:
     - uses: actions/setup-go@v5
       with:
         go-version: 1.21.x
+    - uses: arduino/setup-protoc@v2
```
[the remark this answers is not in the document]
Comment (balajiv113) [reply .github/workflows/release.yml:34]: To install protoc compiler

File: pkg/guestagent/api/const.go (hunk 7 above)
Comment (balajiv113) [reply pkg/guestagent/api/const.go:10]: Will rename the file accordingly

File: .github/workflows/release.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (AkihiroSuda) [reply .github/workflows/release.yml:34]: Can we just use apt-get and brew?

File: Makefile [hunk 10]
```diff
@@ -242,6 +244,18 @@ lint:
 clean:
 	rm -rf _output vendor
 
+.PHONY: gen-proto
+gen-proto:
+	go install google.golang.org/protobuf/cmd/protoc-gen-go@v1.28
+	go install google.golang.org/grpc/cmd/protoc-gen-go-grpc@v1.2
```
[the remark this answers is not in the document]
Comment (balajiv113) [reply Makefile:253]: Separated into install-pkgs

File: Makefile (hunk 1 above)
Comment (balajiv113) [reply Makefile]: Done

File: Makefile (hunk 2 above)
Comment (balajiv113) [reply Makefile]: Done

File: Makefile (hunk 4 above)
Comment (balajiv113) [reply Makefile]: Done

File: pkg/guestagent/api/gen.go (hunk 5 above)
Comment (balajiv113) [reply pkg/guestagent/api/gen.go]: Done

File: pkg/hostagent/hostagent.go (hunk 6 above)
Comment (balajiv113) [reply pkg/hostagent/hostagent.go]: Removed this. Added for debugging but not needed

File: .github/workflows/release.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (balajiv113) [reply .github/workflows/release.yml]: Using apt-get, brew and choco as per os

File: pkg/guestagent/api/const.go (hunk 8 above)
Comment (balajiv113) [reply pkg/guestagent/api/const.go]: Removed this from here and moved to limayaml (This is where its actually used)

File: pkg/guestagent/api/const.go (hunk 7 above)
Comment (balajiv113) [reply pkg/guestagent/api/const.go]: Done

File: Makefile (hunk 3 above)
Comment (balajiv113) [reply Makefile]: Done

File: .github/workflows/release.yml [hunk 11]
```diff
@@ -56,6 +59,10 @@ jobs:
       run: |
         sudo apt-get update
         sudo apt-get install -y gcc-aarch64-linux-gnu
+    - name: Install protoc
+      run: |
+        sudo apt-get update
+        sudo apt-get install -y protobuf-compiler
```
Comment (AkihiroSuda) [inline .github/workflows/release.yml:65]: Should not be needed for `release.yml`

File: .gitignore [hunk 12]
```diff
@@ -1,4 +1,4 @@
 _output/
 _artifacts/
 lima.REJECTED.yaml
-.config
+.config
```
Comment (AkihiroSuda) [inline .gitignore:4]: Seems to be an accidental change

File: Makefile [hunk 13]
```diff
@@ -107,6 +107,8 @@ endif
 
 .PHONY: binaries
 binaries: clean \
+    install-pkgs \
+    generate \
```
Comment (AkihiroSuda) [inline Makefile:111]: `make binaries` should not depend on these targets.

These targets should be only called via CI or via a manual invocation

File: Makefile [hunk 14]
```diff
@@ -242,6 +247,15 @@ lint:
 clean:
 	rm -rf _output vendor
 
+.PHONY: install-pkgs
+install-pkgs:
```
Comment (AkihiroSuda) [inline Makefile:251]: Maybe:
```suggestion
install-tools:
```

File: pkg/guestagent/api/guestservice.pb.desc
Comment (AkihiroSuda) [inline pkg/guestagent/api/guestservice.pb.desc:1]: What's this file?

Comment (balajiv113) [reply pkg/guestagent/api/guestservice.pb.desc:1]: This we rely on for diffing (in lint). In generated code we have protoc version in comments, due to which we cannot check for a proper git diff

File: Makefile (hunk 13 above)
Comment (balajiv113) [reply Makefile:111]: Will make generate depend on install-tools. This way its easy to generate proto without running too many targets

Comment (AkihiroSuda) [reply Makefile:111]: The problem is that this will clover the command binaries installed by the user.

Comment (balajiv113) [reply Makefile:111]: Ah okay got it. So its upto dev/user to run install-tools as well if and when needed

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Comment by afbjorklund** [comment]
I don't want to review the grpc code now, but if it is doesn't break the others I am fine with adding it.

**Comment by AkihiroSuda** [comment]
Can we merge? @lima-vm/maintainers

File: pkg/guestagent/api/server/server.go [hunk 15]
```diff
@@ -2,82 +2,38 @@ package server
...
-	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/types/known/emptypb"
 )
 
-type Backend struct {
-	Agent guestagent.Agent
+func StartServer(lis net.Listener, guest *GuestServer) error {
+	var opts []grpc.ServerOption
+	server := grpc.NewServer(opts...)
```
Comment (alexandear) [inline pkg/guestagent/api/server/server.go:15]: Could we simplify?

```suggestion
	server := grpc.NewServer()
```

File: pkg/guestagent/guestagent_linux.go [hunk 16]
```diff
@@ -162,38 +163,37 @@ func comparePorts(old, neww []api.IPPort) (added, removed []api.IPPort) {
...
-		ev  api.Event
+		ev  = &api.Event{}
 		err error
 	)
 	newSt := st
 	newSt.ports, err = a.LocalPorts(ctx)
 	if err != nil {
 		ev.Errors = append(ev.Errors, err.Error())
-		ev.Time = time.Now()
+		ev.Time = timestamppb.New(time.Now())
```
Comment (alexandear) [inline pkg/guestagent/guestagent_linux.go:175]:
```suggestion
		ev.Time = timestamppb.Now()
```

File: pkg/guestagent/guestagent_linux.go [hunk 17]
```diff
@@ -295,14 +295,14 @@ func (a *agent) LocalPorts(_ context.Context) ([]api.IPPort, error) {
 
 func (a *agent) Info(ctx context.Context) (*api.Info, error) {
 	var (
-		info api.Info
+		info = &api.Info{}
```
Comment (alexandear) [inline pkg/guestagent/guestagent_linux.go:298]: Do we need to change this line?

File: Makefile [hunk 18]
```diff
@@ -231,8 +231,14 @@ uninstall:
 	if [ "$$(readlink "$(DEST)/bin/nerdctl")" = "nerdctl.lima" ]; then rm "$(DEST)/bin/nerdctl"; fi
 	if [ "$$(readlink "$(DEST)/bin/apptainer")" = "apptainer.lima" ]; then rm "$(DEST)/bin/apptainer"; fi
 
+.PHONY: check-generated
+check-generated:
+	@test -z "$$(git status --short | grep ".pb.desc" | tee /dev/stderr)" || \
+        		((git diff $$(find . -name '*.pb.desc') | cat) && \
+        		(echo "Please run 'make generate' when making changes to proto files and check-in the generated file changes" && false))
```
[the remark this answers is not in the document]
Comment (balajiv113) [reply Makefile]: Done

File: pkg/guestagent/api/client/client.go [hunk 19]
```diff
@@ -1,83 +1,52 @@
...
+func NewGuestAgentClient(dialFn func(ctx context.Context) (net.Conn, error)) (*GuestAgentClient, error) {
+	var opts []grpc.DialOption
 
-func (c *client) HTTPClient() *http.Client {
-	return c.Client
-}
+	opts = append(opts, grpc.WithContextDialer(func(ctx context.Context, target string) (net.Conn, error) {
+		return dialFn(ctx)
+	}))
+	opts = append(opts, grpc.WithTransportCredentials(insecure.NewCredentials()))
```
[the remark this answers is not in the document]
Comment (balajiv113) [reply pkg/guestagent/api/client/client.go]: Done

File: pkg/guestagent/api/guestservice.proto [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+import "google/protobuf/empty.proto";
+import "google/protobuf/timestamp.proto";
+
+service GuestService {
+  rpc GetInfo(google.protobuf.Empty) returns (Info);
+  rpc GetEvents(google.protobuf.Empty) returns (stream Event);
+}
+
+message Info {
+  repeated IPPort LocalPorts = 1;
```
[the remark this answers is not in the document]
Comment (balajiv113) [reply pkg/guestagent/api/guestservice.proto]: I have missed it completely. Thanks for pointing out 🙏

File: pkg/guestagent/api/guestservice.proto [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+  google.protobuf.Timestamp Time = 1;
+  repeated IPPort LocalPortsAdded = 2;
+  repeated IPPort LocalPortsRemoved = 3;
+  repeated string Errors = 4;
+}
+
+message IPPort {
+  string IP = 1;
+  int32 Port = 2;
+}
```
[the remark this answers is not in the document]
Comment (balajiv113) [reply pkg/guestagent/api/guestservice.proto]: Done

File: pkg/guestagent/api/server/server.go (hunk 15 above)
Comment (balajiv113) [reply pkg/guestagent/api/server/server.go]: Done

File: pkg/guestagent/api/server/server.go [hunk 22]
```diff
@@ -2,82 +2,38 @@ package server
...
-	go b.Agent.Events(ctx, ch)
-
-	enc := json.NewEncoder(w)
-	for ev := range ch {
-		if err := enc.Encode(ev); err != nil {
-			logrus.Warn(err)
-			return
+func (s GuestServer) GetEvents(_ *emptypb.Empty, stream api.GuestService_GetEventsServer) error {
+	responses := make(chan *api.Event)
+	go s.Agent.Events(context.Background(), responses)
```
[the remark this answers is not in the document]
Comment (balajiv113) [reply pkg/guestagent/api/server/server.go]: Done

File: pkg/guestagent/guestagent_linux.go (hunk 17 above)
Comment (balajiv113) [reply pkg/guestagent/guestagent_linux.go]: Updated

**Review by AkihiroSuda — DISMISSED** [review]
Still LGTM but needs rebase

**Review by alexandear — APPROVED** [review]

**Review by AkihiroSuda — APPROVED** [review]

**Comment by AkihiroSuda** [comment]
```
go generate ./...
 M pkg/guestagent/api/guestservice.pb.desc
diff --git a/pkg/guestagent/api/guestservice.pb.desc b/pkg/guestagent/api/guestservice.pb.desc
index 593bc59..2587c5a 100644
--- a/pkg/guestagent/api/guestservice.pb.desc
+++ b/pkg/guestagent/api/guestservice.pb.desc
@@ -1,9 +1,9 @@
 
-�
+�
 guestservice.protogoogle/protobuf/empty.protogoogle/protobuf/timestamp.proto"0
 Info(
 local_ports (2.IPPortR
make: *** [Makefile:[23](https://github.com/lima-vm/lima/actions/runs/7883027544/job/21509211168?pr=2162#step:6:24)6: check-generated] Error 1
-localPorts"�
+localPorts"�
 Event.
 time (2.google.protobuf.TimestampRtime3
 local_ports_added (2.IPPortRlocalPortsAdded7
Please run 'make generate' when making changes to proto files and check-in the generated file changes
```
https://github.com/lima-vm/lima/actions/runs/7883027544/job/21509211168?pr=2162

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Pull request merged**